storagenode/piecestore/endpoint.go

```diff
@@ -314,7 +334,46 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 	largestOrder := pb.Order{}
 	defer commitOrderToStore(ctx, &largestOrder)
 
+	// Measure upload rate and average upload rate
+	uploadSize := int64(0)
+	previousSize := int64(0)
+
+	// sumUploadRate sums all previously sampled upload rate
+	sumUploadRate := float64(0)
+	// averageUploadRate= sum of sampled upload rate / total of samples
+	averageUploadRate := float64(0)
+	// number as of data chunks that have been received so far
+	// first chunk is counted as 0
+	chunkCount := float64(-1)
+
 	for {
+		// Increment counts
+		chunkCount++
+		// Measure upload rate (bytes per second)
+		currentTime := time.Now().UTC()
+		dt := currentTime.Sub(startTime)
+
+		if pieceWriter != nil {
+			// delta uploaded
+			uploadSize = pieceWriter.Size() - previousSize
+			previousSize = pieceWriter.Size()
+		}
+
+		if dt.Seconds() > 0 {
+			// currentUploadRate mesures the rate of upload for each chunk of data.
+			currentUploadRate := float64(uploadSize) / dt.Seconds()
+			sumUploadRate += currentUploadRate
+
+			// Average upload rate is counted as: total upload rate samples per a period of time
+			// divides by total number of chunks
+			averageUploadRate = sumUploadRate / chunkCount
```

Comment:
- It would not panic, as at moment t0, aka `chunkCount==0`, `sumUploadRate` is also 0.
- However, for explicit, the check is skiped at `chunkCount==0` and also for the first 10 seconds of the upload

Reply:
Ah, I see, `currentUploadRate` is effectively `(pieceWriter.Size() - previousSize) / dt.Seconds()`. I had thought it was `pieceWriter.Size() / dt.Seconds()` instead.

How about just using `currentUploadRate := pieceWriter.Size() / dt.Seconds()`, though? That would be the real average rate over the life time of the transfer.

Reply:
Indeed. That simplifies a lot in the logical thinking as well is in code.

Fixed.